### PR TITLE
fPIC should not always be removed

### PR DESCRIPTION
**Pull request by uilianries**
The Conan Center hooks, validates when fPIC is declared in the recipe, but also requires that fPIC must be removed for Windows. However, some project are not supported on Linux (e.g. libmount), so there is no need for such action.

This PR runs `conan info` to validate the recipe support for Windows. If ConanInvalidConfiguration is raise, it's because the recipe can't be built on Windows, so we don't need to remove fPIC from the recipe. 

Related PR: https://github.com/conan-io/conan-center-index/pull/123

### CODE REVIEW

File: hooks/conan-center.py [hunk 1]
```diff
@@ -155,8 +157,17 @@ def test(out):
                     and any(r in low for r in remove_fpic_option):
                 out.success("OK. 'fPIC' option found and apparently well managed")
             else:
-                out.error("'fPIC' option not managed correctly. Please remove it for Windows "
-                          "configurations: del self.options.fpic")
+                conan_instance, _, _ = conan_api.Conan.factory()
```
Comment (danimtb) [inline hooks/conan-center.py]: I don't like too much the idea os instantiating a Conan API here, but admit is the most robust option... An alternative could be checking for these lines in the recipe:

```
        if self.settings.os == "Windows":
            raise ConanInvalidConfiguration
```

File: tests/test_hooks/conan-center/test_conan-center.py [hunk 2]
```diff
@@ -114,3 +135,16 @@ def test_conanfile_installer(self):
         self.assertIn("ERROR: [PACKAGE LICENSE (KB-H012)] No 'licenses' folder found in package", output)
         self.assertIn("[DEFAULT PACKAGE LAYOUT (KB-H013)] OK", output)
         self.assertIn("[SHARED ARTIFACTS (KB-H015)] OK", output)
+
+    @unittest.skipIf(platform.system() != "Linux", "Only linux")
+    def test_conanfile_linux_only(self):
+        tools.save('conanfile.py', content=self.conanfile_linux_only)
+        output = self.conan(['create', '.', 'linuxonly/version@conan/test'])
+        self.assertIn("[RECIPE METADATA (KB-H003)] OK", output)
```
Comment (danimtb) [inline tests/test_hooks/conan-center/test_conan-center.py]: this check is not related to this PR

File: tests/test_hooks/conan-center/test_conan-center.py [hunk 3]
```diff
@@ -114,3 +135,16 @@ def test_conanfile_installer(self):
...
+        self.assertIn("[FPIC OPTION (KB-H006)] OK", output)
+        self.assertIn("[FPIC MANAGEMENT (KB-H007)] 'fPIC' has not been removed, but this recipe is "
+                      "not supported by Windows.", output)
+        self.assertIn("[LIBCXX MANAGEMENT (KB-H011)] OK", output)
```
Comment (danimtb) [inline tests/test_hooks/conan-center/test_conan-center.py]: same here and the following

File: hooks/conan-center.py (hunk 1 above)
Comment (ericLemanissier) [reply hooks/conan-center.py]: the weakness with this alternative is that sometimes it is `if self.settings.os != "Linux":` (https://github.com/conan-io/conan-center-index/pull/123/files#diff-e89e2904d62404d919e6e127764deb39R22)

Comment (danimtb) [reply hooks/conan-center.py]: yes, you are right. Maybe this is the best alternative we have

Comment (uilianries) [reply hooks/conan-center.py]: yeah, that's why I preferred using Conan API.

**Comment by uilianries** [comment]
> Too much. It will increase the duration of the hook and also will be another maintenance nightmware. Could this check be moved to a `pre_build` where we have the settings and might check only when `conanfile.settings.os == "Windows"`? WDYT?

Sounds good. As the recipe will fail when exporting, pre_build is a safer place to check it.

**Comment by uilianries** [comment]
I've moved fPIC management to post_source, because pre_build doesn't offer conanfile_path.

**Comment by SSE4** [comment]
@uilianries do you have an update? we need this one to fix ODBC build

**Comment by danimtb** [comment]
@uilianries, I think the check could be easier and done in `pre_build()` with something like:

```
if conanfile.settings.os == "Windows" and conanfile.options.fpic:
   raise ...
```
Maybe we have to use `get_safe()` or get the option in another way, but I think that is the idea @lasote was proposing above. It should be more robust that way instead of checking the conanfile content.

**Comment by uilianries** [comment]
@SSE4 @danimtb I think it's better now! I'm using pre_build to check if is windows and fPIC is declared, no more string parser.

File: hooks/conan-center.py [hunk 4]
```diff
@@ -248,6 +230,20 @@ def _is_removing_libcxx():
                         "Can't detect C++ source files but recipe does not remove 'compiler.libcxx'")
 
 
+@raise_if_error_output
+def pre_build(output, conanfile, **kwargs):
+
+    @run_test("KB-H007", output)
+    def test(out):
+        has_fpic = conanfile.options.get_safe("fPIC") or conanfile.options.get_safe("fpic")
```
Comment (ericLemanissier) [inline hooks/conan-center.py]: doesn't `get_safe` return `False` if the option exist and has the value `False` ?

Comment (uilianries) [reply hooks/conan-center.py]: >  EDIT: ok, it doesn't !(sorry for the noise)
np! thanks for reviewing!

Comment (lasote) [inline hooks/conan-center.py]: The option shouldn't be named "fpic". Maybe that check is a `pre-export` one?

Comment (danimtb) [reply hooks/conan-center.py]: I agree, but that is another check that should fall into the syntax side. Let's try to get this PR done and go for that later

Comment (lasote) [reply hooks/conan-center.py]: So, remove the "or" here

Comment (uilianries) [reply hooks/conan-center.py]: done

**Comment by danimtb** [comment]
Merging this as Appveyor CI already passed in previous commits (not affected by the latest ones)
